core/common/src/main/java/alluxio/collections/LockCache.java

```diff
@@ -0,0 +1,248 @@
+      mValue = value;
+      mIsAccessed = false;
+      mRefCount = new AtomicInteger(1);
+    }
+
+    /**
+     * Get the ref counter associated with this value node.
+     *
+     * @return a ref counter
+     */
+    public AtomicInteger getRefCounter() {
+      return mRefCount;
+    }
+
+    /**
+     * Get the value contained in the value node.
+     *
+     * @return the lock contained in the value node
+     */
+    public ReentrantReadWriteLock get() {
+      return mValue;
+    }
+  }
+
+  private static final Logger LOG = LoggerFactory.getLogger(LockCache.class);
+  private static final float DEFAULT_LOAD_FACTOR = 0.75f;
+  static final float SOFT_LIMIT_RATIO = 0.9f;
+
+  private final Map<K, ValNode> mCache;
+
+  /**
+   * SoftLimit = hardLimit * softLimitRatio
+   * once the size reaches softlimit, eviction starts to happen,
+   * once the size reaches hardlimit, blocking starts to happen.
+   */
+  private final int mHardLimit;
+  private final int mSoftLimit;
+  private final Function<? super K, ? extends ReentrantReadWriteLock> mDefaultLoader;
+  private Iterator<Map.Entry<K, ValNode>> mIterator;
+  private final Lock mEvictLock;
+
+  private long mLastLogTime;
+
+  /**
+   * Constructor for a lock cache.
+   *
+   * @param defaultLoader specify a function to generate a value based on a key
+   * @param initialSize initial size of the cache
+   * @param maxSize maximum size of the cache
+   * @param concurrencyLevel concurrency level of the cache
+   */
+  public LockCache(@Nullable Function<? super K, ? extends ReentrantReadWriteLock> defaultLoader,
+      int initialSize, int maxSize, int concurrencyLevel) {
+    mDefaultLoader = defaultLoader;
+
+    mHardLimit = maxSize;
+    mSoftLimit = (int) Math.round(SOFT_LIMIT_RATIO * maxSize);
+    mCache = new ConcurrentHashMap<>(initialSize, DEFAULT_LOAD_FACTOR, concurrencyLevel);
+    mIterator = mCache.entrySet().iterator();
+    mEvictLock = new ReentrantLock();
+    mLastLogTime = System.currentTimeMillis();
+  }
+
+  private void evictIfOverLimit() {
+    long numToEvict = mCache.size() - mSoftLimit;
+
+    if (numToEvict <= 0) {
+      return;
+    }
+
+    if (mEvictLock.tryLock()) {
+      try {
+        // This thread won the race as the evictor
+        while (numToEvict > 0) {
+
+          if (!mIterator.hasNext()) {
+            mIterator = mCache.entrySet().iterator();
+          }
+          Map.Entry<K, ValNode> candidateMapEntry = mIterator.next();
+          ValNode candidate = candidateMapEntry.getValue();
+          if (candidate.mIsAccessed) {
+            candidate.mIsAccessed = false;
+          } else {
+            if (candidate.mRefCount.compareAndSet(0, Integer.MIN_VALUE)) {
+              // the value object can be evicted, at the same time we make refCount minValue
+              mIterator.remove();
+              numToEvict--;
+            }
+          }
+        }
+      } finally {
+        mEvictLock.unlock();
+      }
+    }
+  }
+
+  /**
```

Comment:
this is the most important method from a usage standpoint, can we make it more detailed?

Reply:
good point